// state-control/es/Input/Input.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import PropTypes from 'prop-types';
import { NUMBER_COLOR_TYPE, VALUE_TYPE } from '../common/constants';
import { noop, restoreSelection, saveSelection } from '../common/utils';
import controlled from '../common/controlled';
var defaultNumberColor = '#cfffcf';

var Input = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(Input, _React$Component);

  function Input() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;

    _defineProperty(_assertThisInitialized(_this), "getInnerStyles", function () {
      return {
        display: 'inline-block',
        height: _this.props.multiLine ? '5em' : 'auto',
        verticalAlign: _this.props.multiLine ? 'top' : 'inherit',
        padding: '0.1em 0.2em',
        border: '1px solid darkgrey',
        backgroundColor: function () {
          var _this$props = _this.props,
              readOnly = _this$props.readOnly,
              numberColor = _this$props.numberColor;

          if (readOnly) {
            return '#eee';
          }

          if (numberColor) {
            return numberColor.length ? numberColor : defaultNumberColor;
          }

          return 'white';
        }()
      };
    });

    _defineProperty(_assertThisInitialized(_this), "handleRef", function (cb) {
      return function (element) {
        _this.target = element;
        cb(element);
      };
    });

    _defineProperty(_assertThisInitialized(_this), "eventHandler", function (processor) {
      return function (event) {
        processor(_assertThisInitialized(_this))(event);
        _this.selection = saveSelection(_this.target);
      };
    });

    return _this;
  }

  var _proto = Input.prototype;

  _proto.componentDidUpdate = function componentDidUpdate() {
    restoreSelection(this.target, this.selection);
  };

  _proto.getSnapshotBeforeUpdate = function getSnapshotBeforeUpdate() {
    this.selection = saveSelection(this.target);
    return this.selection;
  };

  _proto.render = function render() {
    var _this$props2 = this.props,
        className = _this$props2.className,
        label = _this$props2.label,
        refHandler = _this$props2.refHandler,
        onFocus = _this$props2.onFocus,
        multiLine = _this$props2.multiLine,
        numberColor = _this$props2.numberColor,
        passedProps = _objectWithoutPropertiesLoose(_this$props2, ["className", "label", "refHandler", "onFocus", "multiLine", "numberColor"]);

    var Inner = multiLine ? 'textarea' : 'input';
    return /*#__PURE__*/React.createElement("div", {
      className: className,
      style: {
        marginBottom: '0.7em'
      }
    }, /*#__PURE__*/React.createElement("label", {
      htmlFor: this.props.id,
      style: {
        display: 'inline-block',
        paddingRight: '0.3em'
      }
    }, label), /*#__PURE__*/React.createElement(Inner, _extends({
      ref: this.handleRef(refHandler(this)),
      onFocus: this.eventHandler(onFocus),
      style: this.getInnerStyles()
    }, passedProps)), /*#__PURE__*/React.createElement("span", {
      style: {
        paddingLeft: '0.3em'
      }
    }, this.props.suffix));
  };

  return Input;
}(React.Component);

_defineProperty(Input, "defaultProps", {
  className: '',
  label: '',
  suffix: '',
  refHandler: noop,
  onClick: noop,
  onFocus: noop,
  onChange: noop,
  value: '',
  multiLine: false,
  readOnly: false,
  numberColor: false
});

Input.propTypes = process.env.NODE_ENV !== "production" ? {
  className: PropTypes.string,
  id: PropTypes.string.isRequired,
  label: PropTypes.string,
  suffix: PropTypes.string,
  refHandler: PropTypes.func,
  onClick: PropTypes.func,
  onFocus: PropTypes.func,
  onChange: PropTypes.func,
  value: VALUE_TYPE,
  multiLine: PropTypes.bool,
  readOnly: PropTypes.bool,
  numberColor: NUMBER_COLOR_TYPE
} : {};
export default controlled(Input);
