// state-control/es/PresetsBlock/PresetsBlock.js
import React from 'react';
import PropTypes from 'prop-types';
import { noop } from '../common/utils';

var Preset = function Preset(_ref) {
  var text = _ref.text,
      tabIndex = _ref.tabIndex,
      onClick = _ref.onClick;
  return /*#__PURE__*/React.createElement("div", {
    style: {
      marginBottom: '0.6em'
    }
  }, /*#__PURE__*/React.createElement("button", {
    type: "button",
    onClick: onClick,
    tabIndex: tabIndex
  }, text));
};

Preset.propTypes = process.env.NODE_ENV !== "production" ? {
  text: PropTypes.string,
  tabIndex: PropTypes.number,
  onClick: PropTypes.func
} : {};
Preset.defaultProps = {
  text: '',
  tabIndex: -1,
  onClick: noop
};

var setParams = function setParams(setHandler, params) {
  return function () {
    Object.entries(params).forEach(function (_ref2) {
      var name = _ref2[0],
          value = _ref2[1];
      setHandler(name, value);
    });
  };
};

export var PresetsBlock = function PresetsBlock(_ref3) {
  var className = _ref3.className,
      setters = _ref3.setters,
      setHandler = _ref3.setHandler,
      tabIndexOffset = _ref3.tabIndexOffset;
  var settersArray = Array.isArray(setters) ? setters.map(function (setter) {
    return [setter.text, setter.params];
  }) : Object.entries(setters);
  return /*#__PURE__*/React.createElement("div", {
    className: className
  }, settersArray.map(function (_ref4, index) {
    var text = _ref4[0],
        params = _ref4[1];
    return /*#__PURE__*/React.createElement(Preset, {
      onClick: setParams(setHandler, params),
      key: text,
      tabIndex: index + tabIndexOffset,
      text: text
    });
  }));
};
PresetsBlock.propTypes = process.env.NODE_ENV !== "production" ? {
  className: PropTypes.string,
  setters: PropTypes.oneOfType([PropTypes.object, PropTypes.array]).isRequired,
  setHandler: PropTypes.func.isRequired,
  tabIndexOffset: PropTypes.number
} : {};
PresetsBlock.defaultProps = {
  className: '',
  tabIndexOffset: 1
};
