// state-control/lib/Check/Check.js
"use strict";

exports.__esModule = true;
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("../common/constants");

var _controlled = _interopRequireDefault(require("../common/controlled"));

var _utils = require("../common/utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Check = /*#__PURE__*/function (_React$PureComponent) {
  _inheritsLoose(Check, _React$PureComponent);

  function Check() {
    return _React$PureComponent.apply(this, arguments) || this;
  }

  var _proto = Check.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        className = _this$props.className,
        value = _this$props.value,
        refHandler = _this$props.refHandler,
        onClick = _this$props.onClick,
        onFocus = _this$props.onFocus,
        label = _this$props.label,
        readOnly = _this$props.readOnly,
        numberColor = _this$props.numberColor,
        type = _this$props.type,
        passedProps = _objectWithoutPropertiesLoose(_this$props, ["className", "value", "refHandler", "onClick", "onFocus", "label", "readOnly", "numberColor", "type"]);

    return /*#__PURE__*/_react["default"].createElement("div", {
      className: className
    }, /*#__PURE__*/_react["default"].createElement("input", _extends({
      type: "checkbox",
      checked: value,
      ref: refHandler(this),
      onClick: onClick(this),
      onFocus: onFocus(this),
      disabled: readOnly,
      style: {
        marginBottom: '0.8em',
        marginLeft: '1px'
      }
    }, passedProps)), /*#__PURE__*/_react["default"].createElement("label", {
      htmlFor: this.props.id
    }, label));
  };

  return Check;
}(_react["default"].PureComponent);

_defineProperty(Check, "defaultProps", {
  className: '',
  label: '',
  value: false,
  refHandler: _utils.noop,
  onClick: _utils.noop,
  onFocus: _utils.noop,
  readOnly: false,
  numberColor: false
});

Check.propTypes = process.env.NODE_ENV !== "production" ? {
  className: _propTypes["default"].string,
  id: _propTypes["default"].string.isRequired,
  label: _propTypes["default"].string,
  value: _propTypes["default"].bool,
  refHandler: _propTypes["default"].func,
  onClick: _propTypes["default"].func,
  onFocus: _propTypes["default"].func,
  readOnly: _propTypes["default"].bool,
  numberColor: _constants.NUMBER_COLOR_TYPE
} : {};

var _default = (0, _controlled["default"])(Check);

exports["default"] = _default;
module.exports = exports.default;
