// state-control/lib/Input/Input.js
"use strict";

exports.__esModule = true;
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("../common/constants");

var _utils = require("../common/utils");

var _controlled = _interopRequireDefault(require("../common/controlled"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var defaultNumberColor = '#cfffcf';

var Input = /*#__PURE__*/function (_React$Component) {
  _inheritsLoose(Input, _React$Component);

  function Input() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _React$Component.call.apply(_React$Component, [this].concat(args)) || this;

    _defineProperty(_assertThisInitialized(_this), "getInnerStyles", function () {
      return {
        display: 'inline-block',
        height: _this.props.multiLine ? '5em' : 'auto',
        verticalAlign: _this.props.multiLine ? 'top' : 'inherit',
        padding: '0.1em 0.2em',
        border: '1px solid darkgrey',
        backgroundColor: function () {
          var _this$props = _this.props,
              readOnly = _this$props.readOnly,
              numberColor = _this$props.numberColor;

          if (readOnly) {
            return '#eee';
          }

          if (numberColor) {
            return numberColor.length ? numberColor : defaultNumberColor;
          }

          return 'white';
        }()
      };
    });

    _defineProperty(_assertThisInitialized(_this), "handleRef", function (cb) {
      return function (element) {
        _this.target = element;
        cb(element);
      };
    });

    _defineProperty(_assertThisInitialized(_this), "eventHandler", function (processor) {
      return function (event) {
        processor(_assertThisInitialized(_this))(event);
        _this.selection = (0, _utils.saveSelection)(_this.target);
      };
    });

    return _this;
  }

  var _proto = Input.prototype;

  _proto.componentDidUpdate = function componentDidUpdate() {
    (0, _utils.restoreSelection)(this.target, this.selection);
  };

  _proto.getSnapshotBeforeUpdate = function getSnapshotBeforeUpdate() {
    this.selection = (0, _utils.saveSelection)(this.target);
    return this.selection;
  };

  _proto.render = function render() {
    var _this$props2 = this.props,
        className = _this$props2.className,
        label = _this$props2.label,
        refHandler = _this$props2.refHandler,
        onFocus = _this$props2.onFocus,
        multiLine = _this$props2.multiLine,
        numberColor = _this$props2.numberColor,
        passedProps = _objectWithoutPropertiesLoose(_this$props2, ["className", "label", "refHandler", "onFocus", "multiLine", "numberColor"]);

    var Inner = multiLine ? 'textarea' : 'input';
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: className,
      style: {
        marginBottom: '0.7em'
      }
    }, /*#__PURE__*/_react["default"].createElement("label", {
      htmlFor: this.props.id,
      style: {
        display: 'inline-block',
        paddingRight: '0.3em'
      }
    }, label), /*#__PURE__*/_react["default"].createElement(Inner, _extends({
      ref: this.handleRef(refHandler(this)),
      onFocus: this.eventHandler(onFocus),
      style: this.getInnerStyles()
    }, passedProps)), /*#__PURE__*/_react["default"].createElement("span", {
      style: {
        paddingLeft: '0.3em'
      }
    }, this.props.suffix));
  };

  return Input;
}(_react["default"].Component);

_defineProperty(Input, "defaultProps", {
  className: '',
  label: '',
  suffix: '',
  refHandler: _utils.noop,
  onClick: _utils.noop,
  onFocus: _utils.noop,
  onChange: _utils.noop,
  value: '',
  multiLine: false,
  readOnly: false,
  numberColor: false
});

Input.propTypes = process.env.NODE_ENV !== "production" ? {
  className: _propTypes["default"].string,
  id: _propTypes["default"].string.isRequired,
  label: _propTypes["default"].string,
  suffix: _propTypes["default"].string,
  refHandler: _propTypes["default"].func,
  onClick: _propTypes["default"].func,
  onFocus: _propTypes["default"].func,
  onChange: _propTypes["default"].func,
  value: _constants.VALUE_TYPE,
  multiLine: _propTypes["default"].bool,
  readOnly: _propTypes["default"].bool,
  numberColor: _constants.NUMBER_COLOR_TYPE
} : {};

var _default = (0, _controlled["default"])(Input);

exports["default"] = _default;
module.exports = exports.default;
