// state-control/lib/PresetsBlock/PresetsBlock.js
"use strict";

exports.__esModule = true;
exports.PresetsBlock = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _utils = require("../common/utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var Preset = function Preset(_ref) {
  var text = _ref.text,
      tabIndex = _ref.tabIndex,
      onClick = _ref.onClick;
  return /*#__PURE__*/_react["default"].createElement("div", {
    style: {
      marginBottom: '0.6em'
    }
  }, /*#__PURE__*/_react["default"].createElement("button", {
    type: "button",
    onClick: onClick,
    tabIndex: tabIndex
  }, text));
};

Preset.propTypes = process.env.NODE_ENV !== "production" ? {
  text: _propTypes["default"].string,
  tabIndex: _propTypes["default"].number,
  onClick: _propTypes["default"].func
} : {};
Preset.defaultProps = {
  text: '',
  tabIndex: -1,
  onClick: _utils.noop
};

var setParams = function setParams(setHandler, params) {
  return function () {
    Object.entries(params).forEach(function (_ref2) {
      var name = _ref2[0],
          value = _ref2[1];
      setHandler(name, value);
    });
  };
};

var PresetsBlock = function PresetsBlock(_ref3) {
  var className = _ref3.className,
      setters = _ref3.setters,
      setHandler = _ref3.setHandler,
      tabIndexOffset = _ref3.tabIndexOffset;
  var settersArray = Array.isArray(setters) ? setters.map(function (setter) {
    return [setter.text, setter.params];
  }) : Object.entries(setters);
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: className
  }, settersArray.map(function (_ref4, index) {
    var text = _ref4[0],
        params = _ref4[1];
    return /*#__PURE__*/_react["default"].createElement(Preset, {
      onClick: setParams(setHandler, params),
      key: text,
      tabIndex: index + tabIndexOffset,
      text: text
    });
  }));
};

exports.PresetsBlock = PresetsBlock;
PresetsBlock.propTypes = process.env.NODE_ENV !== "production" ? {
  className: _propTypes["default"].string,
  setters: _propTypes["default"].oneOfType([_propTypes["default"].object, _propTypes["default"].array]).isRequired,
  setHandler: _propTypes["default"].func.isRequired,
  tabIndexOffset: _propTypes["default"].number
} : {};
PresetsBlock.defaultProps = {
  className: '',
  tabIndexOffset: 1
};
